### PR TITLE
feat(controller): support gcp workload identity federation as auth method for google artifact registry

### LINKED ISSUE #2029: Workload Identity auth to Google Artifact Registry
# Proposed Feature

Kargo should support workload identity auth to GCR.

This issue is the GCR equivalent of https://github.com/akuity/kargo/issues/1885.

# Motivation

Users wish to avoid using long-lived static credentials to GCR.

### PR DESCRIPTION
**Pull request by krancour**
Follow up to #2059 and #2066

Fixes #2029 

This PR makes minor tweaks to ECR support and expands Google Artifact Registry support to include integration with GCP Workload Identity Federation.

### CODE REVIEW

File: internal/credentials/gar/workload_identity_federation.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/patrickmn/go-cache"
+	"google.golang.org/api/iamcredentials/v1"
+
+	"github.com/akuity/kargo/internal/logging"
+)
+
+var garURLRegex = regexp.MustCompile(`^.+\.pkg\.dev/`)
```
Comment (jessesuen) [inline internal/credentials/gar/workload_identity_federation.go:16]: Though google has deprecated Google Container Registry, the `gcr.io` domain is still valid and [is possible to be hosted by artifact registry](https://cloud.google.com/artifact-registry/docs/transition/gcr-on-ar) by existing gcr.io users.

I think this should be relaxed to also check for `gcr.io`

**Review by jessesuen — APPROVED** [review]
Minor nit about valid container repository domain, but other than that LGTM

File: internal/credentials/gar/workload_identity_federation.go (hunk 1 above)
Comment (krancour) [reply internal/credentials/gar/workload_identity_federation.go:16]: Nice catch. Will relax.

Comment (krancour) [reply internal/credentials/gar/workload_identity_federation.go]: Fixed in https://github.com/akuity/kargo/pull/2070/commits/f648029daf0f199c0f193328fac3ccaf27eaee92

**Pull request merged**